apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/codegen/kotlin/adapter/AdapterCommon.kt

```diff
@@ -122,7 +122,7 @@ internal fun readFromResponseCodeBlock(
             } else {
               "null"
             }
-            beginControlFlow("if·(%L.%M($customScalarAdapters.adapterContext.variables(),·$typenameLiteral))", property.condition.codeBlock(), evaluate)
+            beginControlFlow("if·(%L.%M($customScalarAdapters.adapterContext.variables(),·$typenameLiteral,·$customScalarAdapters.adapterContext,·$reader.getPath()))", property.condition.codeBlock(), evaluate)
```

Comment:
With the above comment, we'd ideally have 

```
$customScalarAdapters.adapterContext.variables(),·$typenameLiteral,·$customScalarAdapters.adapterContext.fragmentIds()
```

And it's all nicely symmetrical? I don't think we need `getPath()` here? Better use it when merging the json?

Reply:
We need the path here because that's part of the id of a received fragment. For instance if we're about to read the fields for `ComputerFields` and we are at `computers.2`, we need to know if we received it. We store the ids of the merged fragments at merging time.

Reply:
I thinnk the part I'm missing is when [`AdapterContext.deferredFragmentLabels()`](https://github.com/apollographql/apollo-kotlin/blob/a73c9c06037c8075f41555917e95a3489b4d6445/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/AdapterContext.kt#L31) is going to be called. 

My understanding would be that it would be setup at the beginning of the parsing. Something like (pseudo code):

```kotlin
adapterContext = adapterContext.newBuilder().deferredFragmentLabels(deferredJsonMerger.mergedFragmentIds)).build()

customScalarAdapters = customScalarAdapters.newBuilder().adapterContext(adapterContext).build()

adapter.fromJson(MapJsonReader(deferredJsonMerger.merge), customScalarAdapters)

```

Then you don't the path in the individual "fromJson" since you have a "Map" with additional data (all the fragment ids present in the map)?

Reply:
> yes but it's the current path that you can only get while parsing the JSON - that is why we still need to get it from JsonReader.getPath and pass it to this method. Unless there's a way around it that I'm missing?

To answer this question, I think this fragmentId could be a compile time constant encoded in the BooleanExpression?

Reply:
I think the reason we can't do it without a "dynamic" path is because of arrays.

Remember, we're going to call `fromJson` in a loop (once per payload), and pass a json that contains more fragments each time. 
Let's re-use this example:

**Received payloads:**

```json
// Payload 1
{"data":{"computers":[{"id":"Computer1"},{"id":"Computer2"}]},"hasNext":true}

// Payload 2
{"data":{"cpu":"386","year":1993,"screen":{"resolution":"640x480"}},"path":["computers",0],"hasNext":true}

// Payload 3
{"data":{"cpu":"486","year":1996,"screen":{"resolution":"640x480"}},"path":["computers",1],"hasNext":true}
```

On the 1st item we'll call `fromJson` with a json that looks like this:

```json
{
  "computers": [
    {
      "id": "Computer1"
    },
    {
      "id": "Computer2"
    }
  ]
}
```

and `mergedFragmentIds` = `[]`

On the 2nd item we'll call `fromJson` with a json that looks like this:

```json
{
  "computers": [
    {
      "id": "Computer1",
      "cpu": "386",
      "year": 1993,
      "screen": {
        "resolution": "640x480"
      }
    },
    {
      "id": "Computer2"
    }
  ]
}
```

and `mergedFragmentIds` = `[ {path="computers.0", label=null} ]`

Inside the `Computer` adapter, the code needs to know if it should read the ComputerFields: Yes for path `computers.0`, No for path `computers.1`.

Is there a way I'm not seeing to do it without using the current parsing path?

Reply:
Yes it is! 😅 

In that case do you still think it would be better to rename BLabel? And pass the set of "ids" to evaluate instead of passing the FragmentContext and delegate to it?

Reply:
Yea, no need to rename `BLabel`. I'm still a bit annoyed to add `AdapterContext` as a dependency to `BooleanExpression` but can't really see a way around this so we can refine down the road

Reply:
Yeah I see what you mean. But all right for now, next steps for this PR are:
- replicate for Java codegen
- rebase on top of a branch containing `getPath`
- regenerate the `.expected`

Then I think we can merge it. 

Then, the next PR will actually wire things (use `DeferredJsonMerger` from `HttpNetworkTransport`)

Reply:
Sounds good 👍